# TMDB 5000 영화 DATA SET

# DATA LOAD

In [1]:
import pandas as pd
import numpy as np
import warnings; warnings.filterwarnings('ignore')

movies=pd.read_csv("tmdb_5000_movies.csv")

In [2]:
movies.shape

(4803, 20)

In [3]:
movies.head(2)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   

                                       homepage     id  \
0                   http://www.avatarmovie.com/  19995   
1  http://disney.go.com/disneypictures/pirates/    285   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500

20개의 피처 중 콘텐츠 기반 필터링 추천 분석에 사용할 주요 칼럼만 추출할 것이다.  
  
id, title(영화제목), genres, vote_average(평균 평점), vote_count(평점 투표 수), popularity,  
keywords(영화를 설명하는 주요 키워드 문구), overview(영화에 대한 개요 설명)

In [6]:
movies_df=movies[['id','title','genres','vote_average','vote_count','popularity','keywords','overview']]

In [7]:
pd.set_option('max_colwidth',1000) # dataframe의 셀 크기 증가시킬 때 사용.
movies_df[['genres','keywords']].head(2)

genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]   
1                                          [{"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 28, "name": "Action"}]   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          keywords  
0  [{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "space war"}, {"id": 3388, "name": "space colony"}, {"id": 3679, "name": "society"}, {"id": 3801, "name": "space travel"}, {"id": 9685, "name": "futuristic"}, {"id": 9840, "name": "romance"}, {"id": 9882, "name": "space"}, {"id": 9951, "name": "alien"}, {"id": 10148, "name": "tribe"}, {"id": 10158, "name": "alien planet"}, {"id": 10987, "name": "cgi"}, {"id": 11399, "name": "marine"}, {"id": 13065, "name": "soldier"}, {"id": 14643, "name": "battle"}, {"id": 14720, "name": "love affair"}, {"id": 165431, "name": "anti war"}, {"id": 193554, "name": "power relations"}, {"id": 206690, "name": "mind and soul"}, {"id": 209714, "name": "3d"}]  
1                                                                                                                                                   [{"id": 270, "name": "ocean"}, {"id": 726, "name": "drug abuse"}, {"id": 911, "name": "exotic island"}, {"id": 1319, "name": "east india trading company"}, {"id": 2038, "name": "love of one's life"}, {"id": 2052, "name": "traitor"}, {"id": 2580, "name": "shipwreck"}, {"id": 2660, "name": "strong woman"}, {"id": 3799, "name": "ship"}, {"id": 5740, "name": "alliance"}, {"id": 5941, "name": "calypso"}, {"id": 6155, "name": "afterlife"}, {"id": 6211, "name": "fighter"}, {"id": 12988, "name": "pirate"}, {"id": 157186, "name": "swashbuckler"}, {"id": 179430, "name": "aftercreditsstinger"}]

보는 것과 같이 genre는  여러 개의 개별 데이터를 갖고 있으며 keywords 또한 마찬가지다.  
dictionary 형태로 되어있는 이것을 key인 'name'으로 추출할 수 있다.

In [8]:
from ast import literal_eval
movies_df['genres']=movies_df['genres'].apply(literal_eval)
movies_df['keywords']=movies_df['keywords'].apply(literal_eval)

ast 모듈의 literal_eval 함수를 이용하면 문자열이 의미하는 list[dict1,dict2]객체로 만들 수 있다!! 

In [9]:
movies_df['genres'][0]

[{'id': 28, 'name': 'Action'},
 {'id': 12, 'name': 'Adventure'},
 {'id': 14, 'name': 'Fantasy'},
 {'id': 878, 'name': 'Science Fiction'}]

In [10]:
movies_df['genres']=movies_df['genres'].apply(lambda x : [y['name'] for y in x])
movies_df['keywords']=movies_df['keywords'].apply(lambda x : [y['name'] for y in x])

In [11]:
movies_df[['genres','keywords']].head(3)

genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   

                                                                                                                                                                                                                   keywords  
0  [culture clash, future, space war, space colony, society, space travel, futuristic, romance, space, alien, tribe, alien planet, cgi, marine, soldier, battle, love affair, anti war, power relations, mind and soul, 3d]  
1              [ocean, drug abuse, exotic island, east india trading company, love of one's life, traitor, shipwreck, strong woman, ship, alliance, calypso, afterlife, fighter, pirate, swashbuckler, aftercreditsstinger]  
2                                                                                                                                  [spy, based on novel, secret agent, sequel, mi6, british secret service, united kingdom]

## 장르 콘텐츠 유사도 측정

## CountVectorizer

가장 간단한 방법인 genres를 문자열로 변경한 뒤 이를 CountVectorizer로 피처벡터화한 행렬 데이터 값을 코사인 유사도로 비교를 시도  

그래서 먼저 CountVectorizer와 관련 개념에 대해 알아보자!!

+ BOW(Bag Of Words) --> 문서를 숫자 벡터로 변환하는 가장 기본적인 방법이다. BOW 방법에서는 전체 문서  {d1,d2,…,dn}  를 구성하는 고정된 단어장(vocabulary)  {t1,t2,…,tm}  를 만들고  di 라는 개별 문서에 단어장에 해당하는 단어들이 포함되어 있는지를 표시하는 방법이다.

   xi,j=문서 di내의 단어 tj의 출현 빈도
 
   또는 xi,j=만약 단어 tj가 문서 di 안에 없으면 0, 만약 단어 tj가 문서 di 안에 있으면 1

+ CountVectorizer --> 문서 집합에서 단어 토큰을 생성하고 각 단어의 수를 세어 BOW 인코딩한 벡터를 만든다.

   CountVectorizer 과정  
   
   1.문서를 토큰 리스트로 변환한다.  
   
   2.각 문서에서 토큰의 출현 빈도를 센다.  
   
   3.각 문서를 BOW 인코딩 벡터로 변환한다.


한번 그렇다면 이것을 예시로 파악해보자

In [12]:
from sklearn.feature_extraction.text import CountVectorizer
corpus = [
    'This is the first document.',
    'This is the second second document.',
    'And the third one.',
    'Is this the first document?',
    'The last document?',
]
vect = CountVectorizer()
vect.fit(corpus)
vect.vocabulary_

{'this': 9,
 'is': 3,
 'the': 7,
 'first': 2,
 'document': 1,
 'second': 6,
 'and': 0,
 'third': 8,
 'one': 5,
 'last': 4}

In [13]:
vect.transform(['This is the second document.']).toarray()

array([[0, 1, 0, 1, 0, 0, 1, 1, 0, 1]], dtype=int64)

This is the second document 라는 문장을 적용하여 나온 array 결과이다.

In [14]:
vect.transform(['Something completely new.']).toarray()

array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=int64)

Something completely new 라는 문장을 적용하여 나온 array 결과이다. 이번에는 어떠한 토큰도 겹치지 않아 array의 결과가 모두 0으로 나온 것이 확인된다.

In [15]:
vect.transform(corpus).toarray()

array([[0, 1, 1, 1, 0, 0, 0, 1, 0, 1],
       [0, 1, 0, 1, 0, 0, 2, 1, 0, 1],
       [1, 0, 0, 0, 0, 1, 0, 1, 1, 0],
       [0, 1, 1, 1, 0, 0, 0, 1, 0, 1],
       [0, 1, 0, 0, 1, 0, 0, 1, 0, 0]], dtype=int64)

vect 라는 CountVectorizer에 fit 한 결과 다음과 같은 array 결과가 나오게 된다.  
  
  그렇다면 이제 실제로 적용해보자

In [16]:
from sklearn.feature_extraction.text import CountVectorizer

# CountVectorizer를 적용하기 위해 공백문자로 word 단위가 구분되는 문자열로 반환.
movies_df['genres_literal']=movies_df['genres'].apply(lambda x : (' ').join(x))

In [17]:
movies_df['genres_literal']

0       Action Adventure Fantasy Science Fiction
1                       Adventure Fantasy Action
2                         Action Adventure Crime
3                    Action Crime Drama Thriller
4               Action Adventure Science Fiction
                          ...                   
4798                       Action Crime Thriller
4799                              Comedy Romance
4800               Comedy Drama Romance TV Movie
4801                                            
4802                                 Documentary
Name: genres_literal, Length: 4803, dtype: object

In [18]:
count_vect=CountVectorizer(min_df=0,ngram_range=(1,2))
genre_mat=count_vect.fit_transform(movies_df['genres_literal'])
print(genre_mat.shape)

(4803, 276)


여기서 CountVectorizer를 생성할 때 min_df와 ngram_range를 설정하였다. 이것은 어떤 것일까?
+ min_df: 단어장에 포함될 최소의 빈도이다. 따라서 0으로 설정한 것은 0의 빈도도 결과로서 활용한다는 것이다.
+ ngram_range: n-gram이라는 것은 단어의 묶음을 말합니다.
이 단어의 묶음을 범위를 설정하는 것이 ngram_range 파라미터입니다.

예를들어, ngram_range = (1, 1)이라면 단어의 묶음을 1개부터 1개까지 설정하라는 의미이다.

만약 ngram_range = (1, 2)라고 한다면, 단어의 묶음을 1개부터 2개까지 설정하라는 뜻이다.
단어사전에는 1개 단어묶음도 있고, 2개 단어묶음도 존재하게 되는것이다.
가령, 'go home', 'very good'과 같은 2개 짜리 묶음도 인덱스를 받게 되는 것이다.

## 코사인 유사도 계산

그렇다면 이제 4803개의 record와 276개의 개별 단어 피처로 구성된 피처 행렬 벡터를 cosine_similarity()를 이용해 코사인 유사도를 계산해보자!!!

In [19]:
from sklearn.metrics.pairwise import cosine_similarity

genre_sim=cosine_similarity(genre_mat, genre_mat)
print(genre_sim.shape)
print(genre_sim[:1])

(4803, 4803)
[[1.         0.59628479 0.4472136  ... 0.         0.         0.        ]]


In [20]:
genre_sim_sorted_ind=genre_sim.argsort()[:,::-1]
print(genre_sim_sorted_ind[:1])

[[   0 3494  813 ... 3038 3037 2401]]


반환된 것을 보면 0번 레코드의 경우 자신인 0번 레코드를 제외하면 3494번 레코드가 가장 유사도가 높고, 그다음이 813번 레코드 그리고 가장 유사도가 낮은 레코드는 2401 레코드라는 뜻이다.

## 장르 콘텐츠 필터링을 이용한 영화 추천

이제 장르 유사도에 따라 영화를 추천하는 함수를 생성하겠다.  
- 함수명은 find_sim_movie()  
- 기반 데이터는 movied_df  
- 레코드별 장르 코사인 유사도 인덱스를 갖고 있는 genre_sim_sorted_ind(sorted_ind)
- 고객이 선정한 추천 기준이 되는 영화 제목  
- 추천할 영화 건수  
  
  위와 같이 입력하면 추천 영화 정보를 가지는 DataFrame을 반환해보자



In [23]:
def find_sim_movie(df,sorted_ind,title_name,top_n=10):
    # 인자로 입력된 movies_df DataFrame에서 'title' 칼럼이 입력된 title_name 값인 DataFrame 추출
    title_movie=df[df['title']==title_name]
    
    # title_named을 가진 DataFrame의 index 객체를 ndarray로 반환하고
    # sorted_ind 인자로 입력된 genre_sim_sorted_ind 객체에서 유사도 순으로 top_n개의 index 추출
    title_index=title_movie.index.values
    similar_indexes=sorted_ind[title_index,:(top_n)]
    
    
    # 추출된 top_n index 출력. top_n index는 2차원 데이터임.
    # dataframe에서 index로 사용하기 위해서 1차원 array로 변경
    print(similar_indexes)
    similar_indexes=similar_indexes.reshape(-1)
    
    return df.iloc[similar_indexes]
    
    

In [24]:
similar_movies=find_sim_movie(movies_df,genre_sim_sorted_ind,'The Godfather',10)

[[2731 1243 3636 1946 2640 4065 1847 4217  883 3866]]


In [25]:
similar_movies[['title','vote_average']]

title  vote_average
2731                          The Godfather: Part II           8.3
1243                                    Mean Streets           7.2
3636                                   Light Sleeper           5.7
1946  The Bad Lieutenant: Port of Call - New Orleans           6.0
2640         Things to Do in Denver When You're Dead           6.7
4065                                      Mi America           0.0
1847                                      GoodFellas           8.2
4217                                            Kids           6.8
883                              Catch Me If You Can           7.7
3866                                     City of God           8.1

대부 2편이 가장 먼저 추천된 것을 확인할 수 있다. 그 외에 1874번 인덱스의 'GoodFellas'도 대부와 비슷한 유형.  
하지만 Light Sleeper, Mi America, Kids 등 대부를 좋아하는 고객에게 섣불리 추천하기에는 이해하기 어려운 영화도 존재한다.  
따라서 좀 더 개선을 한 추천이 필요하다.

그래서 이번에는 일단 좀 더 많은 후보군을 선정한 뒤에 영화의 평점에 따라 필터링해서 최종 추천하는 방식으로 해보겠다.  
영화의 평점 정보인 'vote_average'를 내림차순으로 정렬하여 말이다.

In [26]:
movies_df[['title','vote_average','vote_count']].sort_values('vote_average',ascending=False)[:10]

title  vote_average  vote_count
3519          Stiff Upper Lips          10.0           1
4247     Me You and Five Bucks          10.0           2
4045     Dancer, Texas Pop. 81          10.0           1
4662            Little Big Top          10.0           1
3992                 Sardaarji           9.5           2
2386            One Man's Hero           9.3           2
2970        There Goes My Baby           8.5           2
1881  The Shawshank Redemption           8.5        8205
2796     The Prisoner of Zenda           8.4          11
3337             The Godfather           8.4        5893

하지만, 위의 결과를 보면 vote_count도 고려해야 한다는 것을 확인할 수 있다. 소수의 평가로 고평가 되어지는 영화가 존재하기 때문이다.  
따라서 유명한 영화 평점 사이트인 IMDB에서 사용하는 vote_count를 고려한 평점 방식을 사용하자.  

가중 평점=(v/(v+m)) * R + (m/(v+m)) * C  
+ v: 개별 영화에 평점을 투표한 횟수 == vote_count
+ m: 평점을 부여하기 위한 최소 투표 횟수
+ R: 개별 영화에 대한 평균 평점 == vote_average
+ C: 전체 영화에 대한 평균 평점 == movies_df['vote_average'].mean()

In [29]:
C=movies_df['vote_average'].mean()
#quntile() 상위 ~%까지 추출
m=movies_df['vote_count'].quantile(0.6)
print('C:',round(C,3),'m:',round(m,3))


C: 6.092 m: 370.2


기존 평점을 새로운 가중 평점으로 변경하는 함수를 생성하고 이를 이용해 새로운 평점 정보인 'vote_weighted' 값을 만들겠다.

In [45]:
percentile=0.6
m=movies['vote_count'].quantile(percentile)
C=movies['vote_average'].mean()

def weighted_vote_average(record):
    v=record['vote_count']
    R=record['vote_average']
    
    return ((v/(v+m)) * R + ((m/(m+v)) * C))

movies_df['weighted_vote']=movies.apply(weighted_vote_average,axis=1)

In [48]:
movies_df[['title','vote_average','weighted_vote','vote_count']].sort_values('weighted_vote',ascending=False)[:10]

title  vote_average  weighted_vote  vote_count
1881  The Shawshank Redemption           8.5       8.396052        8205
3337             The Godfather           8.4       8.263591        5893
662                 Fight Club           8.3       8.216455        9413
3232              Pulp Fiction           8.3       8.207102        8428
65             The Dark Knight           8.2       8.136930       12002
1818          Schindler's List           8.3       8.126069        4329
3865                  Whiplash           8.3       8.123248        4254
809               Forrest Gump           8.2       8.105954        7927
2294             Spirited Away           8.3       8.105867        3840
2731    The Godfather: Part II           8.3       8.079586        3338

새롭게 부여된 weighted_vote 평점이 높은 순으로 상위 10개의 영화를 추출한 결과이다.  
이렇게 새롭게 정의된 평점 기준에 따라서 영화를 추천해 볼 것이다.  

장르 유사성이 높은 영화를 top_n의 2배수만큼 후보군으로 선정한 뒤에 weighted_vote 칼럼 값이 높은 순으로  
top_n만큼 추출하는 방식으로 find_sim_movie() 함수를 변경한다.

In [52]:
def find_sim_movie(df,sorted_ind,title_name,top_n=10):
    title_movie=df[df['title']==title_name]
    title_index=title_movie.index.values
    
    #top_n의 2배에 해당하는 장르 유사성이 높은 인덱스 추출
    similar_indexes= sorted_ind[title_index, :(top_n*2)]
    similar_indexes= similar_indexes.reshape(-1)
    
    # 기준 영화 인덱스는 제외
    similar_indexes= similar_indexes[similar_indexes !=title_index]
    
    # top_n의 2배에 해당하는 후보군에서 weighted_vote가 높은 순으로 top_n만큼 추출
    return df.iloc[similar_indexes].sort_values('weighted_vote',ascending=False)[:top_n]

similar_movies=find_sim_movie(movies_df,genre_sim_sorted_ind,'The Godfather',10)
similar_movies[['title', 'vote_average','weighted_vote']]

title  vote_average  weighted_vote
2731       The Godfather: Part II           8.3       8.079586
1847                   GoodFellas           8.2       7.976937
3866                  City of God           8.1       7.759693
1663  Once Upon a Time in America           8.2       7.657811
883           Catch Me If You Can           7.7       7.557097
281             American Gangster           7.4       7.141396
4041              This Is England           7.4       6.739664
1149              American Hustle           6.8       6.717525
1243                 Mean Streets           7.2       6.626569
2839                     Rounders           6.9       6.530427